### PR TITLE
Making mysql and postgres test suites adaptable

### PR DESCRIPTION
**Pull request by karllhughes**
My first shot at #194 

This allows the test classes for Postgres and MySQL to look at their versions and extensions to determine which SQL file to seed the database with. Then, during the test run, it will skip a test that is not supported by the DB version.

This allowed me to run the test suite with the following Postgres and MySQL database/extension combos:

- Postgres 9.1 w/ gis
- Postgres 9.2 w/ gis
- Postgres 9.3 w/ gis
- Postgres 9.4 w/ gis
- Postgres 9.5 w/ gis
- Postgres 9.1 no gis
- Postgres 9.2 no gis
- Postgres 9.3 no gis
- Postgres 9.4 no gis
- Postgres 9.5 no gis
- MySQL 5.5
- MySQL 5.6
- MySQL 5.7

I could not get the test suite to run on Postgres 9.0:

- Postgres 9.0 w/ gis - CREATE EXTENSION SQL command doesn't work
- Postgres 9.0 no gis - relation "pg_extension" does not exist

So that's something we could come back to later.

SQLite and SQL Server tests are unchanged.

### CODE REVIEW

File: tests/PostgresqlTest.php [hunk 1]
```diff
@@ -36,16 +37,123 @@ public function seedDatabase()
...
+
+    public function testWriteProductProperties()
+    {
+        if (version_compare(static::$pg_server_version, '9.2.0') < 0) {
+            $this->markTestSkipped("Postgres < 9.2.0 does not support JSON fields.");
+        }
+        parent::testWriteProductProperties();
+    }
+
+    public function testListProducts()
```
Comment (karllhughes) [inline tests/PostgresqlTest.php]: These empty tests that simply call their parent allow us to keep the order of tests the same when it matters. As I mentioned in #242 I'd like to eventually help refactor the tests to run in greater isolation, but for now this is my hacky workaround 😄

File: tests/data/blog_mysql_55.sql [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+SET foreign_key_checks = 0;
+SET sql_mode = 'NO_AUTO_VALUE_ON_ZERO';
+
+DROP TABLE IF EXISTS `categories`;
+CREATE TABLE `categories` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `name` varchar(255) NOT NULL,
+  `icon` blob NULL,
+  PRIMARY KEY (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci;
```
Comment (karllhughes) [inline tests/data/blog_mysql_55.sql]: Each of the database files is very similar, but has a few lines modified depending on the version being run.

We could probably improve this by making the SQL file generation happen in code, but that adds a whole new level of complexity so I didn't tackle it yet.

File: tests/MysqlTest.php [hunk 3]
```diff
@@ -34,16 +36,138 @@ public function seedDatabase()
...
     }
+
+    /**
+     * Gets the path to the seed file based on the version of MySQL
+     *
+     * @return string
+     */
+    protected static function getSeedFile()
+    {
+        if (static::$mysql_version >= self::MYSQL_57) {
```
Comment (mevdschee) [inline tests/MysqlTest.php]: This should probably also detect MariaDB versions 5.5, 10.0, 10.1 and 10.2

File: tests/PostgresqlTest.php [hunk 4]
```diff
@@ -4,6 +4,9 @@
 
 class PostgresqlTest extends PHP_CRUD_API_Test
 {
+    public static $gis_installed;
```
Comment (mevdschee) [inline tests/PostgresqlTest.php]: A variable named $gis_available would be more generic, or even a $features array containing 'gis' as a value.

File: tests/data/blog_mysql_55.sql (hunk 2 above)
Comment (mevdschee) [inline tests/data/blog_mysql_55.sql]: maybe we can make the fixture add text fields and then convert them to either gis or json after loading the fixture (based on the version/feature detection) in the test suite.

File: tests/MysqlTest.php (hunk 3 above)
Comment (karllhughes) [inline tests/MysqlTest.php]: Ah, good call. I forgot about Mariadb. I'll look at adding that to the mix as well.

**Comment by karllhughes** [comment]
Thanks for the feedback @mevdschee! I'll revise and let you know when it's ready for another look.

**Comment by mevdschee** [comment]
>I'll revise and let you know when it's ready for another look.

That would be great!

>I could not get the test suite to run on Postgres 9.0

I don't think we should aim for anything below 9.1

**Comment by mevdschee** [comment]
Thank you for your contribution! I used your code to implement this feature.

**Comment by karllhughes** [comment]
@mevdschee cool, thanks for letting me know! I was going to make the updates you suggested, I've just been very busy this week. 

I'll look into the failing test and see if I can fix it tonight.

**Comment by karllhughes** [comment]
@mevdschee I love what you did to clean this up, by the way! Your solution looks much cleaner than mine. Clean tests always make me feel good 🛁

**Comment by mevdschee** [comment]
@karllhughes Thank you for your kind words. I also got to play with TravisCI and have installed their docker images locally. Unfortunately that got me nowhere (I got stuck) and I realized how difficult it is to set up these TravisCI tests. 

So now I'm reinstalling and documenting my VM's, so that I can ensure the code quality.

see: https://github.com/mevdschee/php-crud-api/wiki/Home

**Comment by karllhughes** [comment]
Good idea. You could probably start cataloguing more of the common issues in the Wiki too. I notice a lot of "Help, I can't get this set up" type questions coming in on Github.

It's great that so many people find this useful though!

**Comment by mevdschee** [comment]
>It's great that so many people find this useful though!

Yes and (not unimportant) it is useful to myself as well.

**Pull request merged**